Position-Specific Scoring Matrix ([PSSM](https://en.wikipedia.org/wiki/Position_weight_matrix)) is a matrix used to score alignments between a protein sequence (cf. DNA and RNA) and a profile, which is a representation of a multiple sequence alignment. Each score indicates the likelihood of a specific nucleotide or amino acid occurring at a particular position in the sequence. There are many models to calculate it.

### Blast
The Basic local alignment search tool (Blast) [^1] can be used to generate one kind of the PSSM. It can be accessed like that below.

[^1]: Altschul, S.F., Gish, W., Miller, W., Myers, E.W., Lipman, D.J. (1990) “Basic local alignment search tool.” J. Mol. Biol. 215:403-410. https://doi.org/10.1016/S0022-2836(05)80360-2

:material-language-python: Python
``` py linenums="1"
import pypropel as pp

aac = pp.fpmsa.pssm(
    fpn=to('data/pssm/1aigL.pssm'),
    mode='blast',
)
```

:material-note-multiple-outline: Output
``` shell
{1: [-5.0, 1.0, 3.0, 2.0, 3.0, 1.0, 3.0, 2.0, 2.0, 3.0, 2.0, 2.0, 2.0, 2.0, 2.0, -1.0, -1.0, 1.0, 4.0, 3.0], 2: [2.0, 2.0, 4.0, 3.0, 1.0, 2.0, 3.0, -2.0, 2.0, -3.0, -6.0, 3.0, 3.0, 1.0, 1.0, 2.0, -0.0, -1.0, 3.0, 2.0], ...}
```

### HHM

HHM is created by [profile hidden Markov models (HMMs)](https://www.ebi.ac.uk/training/online/courses/pfam-creating-protein-families/what-are-profile-hidden-markov-models-hmms/). The `.hhm` file is generated by HHblits[^2]. It measures PSSM about how conserved each amino acid is at each column in a MSA.

[^2]: Remmert M, Biegert A, Hauser A, Söding J. HHblits: lightning-fast iterative protein sequence searching by HMM-HMM alignment. Nat Methods. 2011 Dec 25;9(2):173-5. doi: 10.1038/nmeth.1818. PMID: 22198341.


:material-language-python: Python
``` py linenums="1"
import pypropel as pp

aac = pp.fpmsa.pssm(
    fpn=to('data/hhm/1aigL.hhm'),
    mode='hhm',
)
```

:material-note-multiple-outline: Output
``` shell
{1: array([1. , 0. , 0. , 0. , 0. , 0. , 0. , 0. , 0. , 0. , 0. , 0. , 0. ,
       0. , 0. , 0. , 0. , 0. , 0. , 0. , 1. , 0. , 0. , 0. , 0. , 0. ,
       0. , 0.5, 1. , 1. ]), 2: array([0. , 0. , 0. , 0. , 0. , 0. , 0. , 0. , 0. , 1. , 0. , 0. , 0. ,
       0. , 0. , 0. , 0. , 0. , 0. , 0. , 1. , 0. , 0. , 0. , 0. , 0. ,
       0. , 0.5, 1. , 1. ]), ...}
```


### Evolutionary profile
Evolutionary profile is calculated as frequencies of amino acids at each MSA column [^2].

[^2]: Hönigschmid P, Frishman D. Accurate prediction of helix interactions and residue contacts in membrane proteins. J Struct Biol. 2016 Apr;194(1):112-23. doi: 10.1016/j.jsb.2016.02.005. Epub 2016 Feb 3. PMID: 26851352.

:material-language-python: Python
``` py linenums="1"
import pypropel as pp

ep = pp.fpmsa.pssm(
    msa=msa,
    mode='ep',
)
```

:material-note-multiple-outline: Output
``` shell
{'A': array([8.25272552e-01, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 5.30380817e-03, 0.00000000e+00,
       0.00000000e+00, 1.06076163e-03, ..., 3.18228490e-03, 1.16683780e-02,
       1.06076163e-03]), 'C': array([0.00000000e+00, ...
```

You can pass `ep_norm` onto `mode` to get the normalised evolutionary profile.

``` py linenums="1"
mode='ep_norm'
```